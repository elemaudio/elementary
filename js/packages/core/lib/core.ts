import {
  createNode,
  isNode,
  resolve,
} from '../nodeUtils';

import type {ElemNode, NodeRepr_t} from '../nodeUtils';


// Generic
type OptionalKeyProps = {
  key?: string,
};

// Const node and constant value nodes
type ConstNodeProps = {
  key?: string,
  value: number,
};

export function constant(props: ConstNodeProps): NodeRepr_t {
  return createNode("const", props, []);
}

export function sr(): NodeRepr_t {
  return createNode("sr", {}, []);
}

export function time(): NodeRepr_t {
  return createNode("time", {}, []);
}

// Counter node
export function counter(gate: ElemNode): NodeRepr_t;
export function counter(props: OptionalKeyProps, gate: ElemNode): NodeRepr_t;
export function counter(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("counter", {}, [resolve(a)]);
  }

  return createNode("counter", a, [resolve(b)]);
}

// Counter2 node
export function counter2(gate: ElemNode): NodeRepr_t;
export function counter2(props: OptionalKeyProps, gate: ElemNode): NodeRepr_t;
export function counter2(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("counter2", {}, [resolve(a)]);
  }

  return createNode("counter2", a, [resolve(b)]);
}

// Accum node
export function accum(xn: ElemNode, reset: ElemNode): NodeRepr_t;
export function accum(props: OptionalKeyProps, xn: ElemNode, reset: ElemNode): NodeRepr_t;
export function accum(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("accum", {}, [resolve(a), resolve(b)]);
  }

  return createNode("accum", a, [resolve(b), resolve(c)]);
}

// Phasor node
export function phasor(rate: ElemNode): NodeRepr_t;
export function phasor(props: OptionalKeyProps, rate: ElemNode): NodeRepr_t;
export function phasor(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("phasor", {}, [resolve(a)]);
  }

  return createNode("phasor", a, [resolve(b)]);
}

export function syncphasor(rate: ElemNode, reset: ElemNode): NodeRepr_t;
export function syncphasor(props: OptionalKeyProps, rate: ElemNode, reset: ElemNode): NodeRepr_t;
export function syncphasor(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("sphasor", {}, [resolve(a), resolve(b)]);
  }

  return createNode("sphasor", a, [resolve(b), resolve(c)]);
}

// Latch node
export function latch(t: ElemNode, x: ElemNode): NodeRepr_t;
export function latch(props: OptionalKeyProps, t: ElemNode, x: ElemNode): NodeRepr_t;
export function latch(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("latch", {}, [resolve(a), resolve(b)]);
  }

  return createNode("latch", a, [resolve(b), resolve(c)]);
}

// MaxHold node
type MaxHoldNodeProps = {
  key?: string,
  hold?: number,
};

export function maxhold(x: ElemNode, reset: ElemNode): NodeRepr_t;
export function maxhold(props: MaxHoldNodeProps, x: ElemNode, reset: ElemNode): NodeRepr_t;
export function maxhold(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("maxhold", {}, [resolve(a), resolve(b)]);
  }

  return createNode("maxhold", a, [resolve(b), resolve(c)]);
}

// Once node
type OnceNodeProps = {
  key?: string,
  arm?: boolean,
};

export function once(x: ElemNode): NodeRepr_t;
export function once(props: OnceNodeProps, x: ElemNode): NodeRepr_t;
export function once(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("once", {}, [resolve(a)]);
  }

  return createNode("once", a, [resolve(b)]);
}

// Rand node
type RandNodeProps = {
  key?: string,
  seed?: number,
};

export function rand(): NodeRepr_t;
export function rand(props: RandNodeProps): NodeRepr_t;
export function rand(a?) {
  if (typeof a !== "undefined") {
    return createNode("rand", a, []);
  }

  return createNode("rand", {}, []);
}

// Metro node
type MetroNodeProps = {
  key?: string,
  interval?: number,
};

export function metro(): NodeRepr_t;
export function metro(props: MetroNodeProps): NodeRepr_t;
export function metro(a?) {
  if (typeof a !== "undefined") {
    return createNode("metro", a, []);
  }

  return createNode("metro", {}, []);
}

// Sample node
type SampleNodeProps = {
  key?: string,
  path?: string,
  mode?: string,
  startOffset?: number,
  stopOffset?: number,
};

export function sample(props: SampleNodeProps, trigger: ElemNode, rate: ElemNode): NodeRepr_t {
  return createNode("sample", props, [resolve(trigger), resolve(rate)]);
}

// Table node
type TableNodeProps = {
  key?: string,
  path?: string,
};

export function table(props: TableNodeProps, t: ElemNode): NodeRepr_t {
  return createNode("table", props, [resolve(t)]);
}

// Convolve node
type ConvolveNodeProps = {
  key?: string,
  path?: string,
};

export function convolve(props: ConvolveNodeProps, x: ElemNode): NodeRepr_t {
  return createNode("convolve", props, [resolve(x)]);
}

// Seq node
type SeqNodeProps = {
  key?: string,
  seq?: Array<number>,
  offset?: number,
  hold?: boolean,
  loop?: boolean,
};

export function seq(props: SeqNodeProps, trigger: ElemNode, reset: ElemNode): NodeRepr_t {
  return createNode("seq", props, [resolve(trigger), resolve(reset)]);
}

// Seq2 node
type Seq2NodeProps = {
  key?: string,
  seq?: Array<number>,
  offset?: number,
  hold?: boolean,
  loop?: boolean,
};

export function seq2(props: Seq2NodeProps, trigger: ElemNode, reset: ElemNode): NodeRepr_t {
  return createNode("seq2", props, [resolve(trigger), resolve(reset)]);
}

// SparSeq node
type SparSeqNodeProps = {
  key?: string,
  seq?: Array<{value: number, tickTime: number}>,
  offset?: number,
  loop?: boolean | Array<number>,
  resetOnLoop?: boolean,
  interpolate?: number,
  tickInterval?: number,
};

export function sparseq(props: SparSeqNodeProps, trigger: ElemNode, reset: ElemNode): NodeRepr_t {
  return createNode("sparseq", props, [resolve(trigger), resolve(reset)]);
}

// SparSeq2 node
type SparSeq2NodeProps = {
  key?: string,
  seq?: Array<{value: number, time: number}>,
};

export function sparseq2(props: SparSeq2NodeProps, time: ElemNode): NodeRepr_t {
  return createNode("sparseq2", props, [resolve(time)]);
}

// Sampleseq node
type SampleSeqNodeProps = {
  key?: string,
  seq?: Array<{value: number, time: number}>,
  duration: number,
  path: string,
};

export function sampleseq(props: SampleSeqNodeProps, time: ElemNode): NodeRepr_t {
  return createNode("sampleseq", props, [resolve(time)]);
}

// Sampleseq2 node
type SampleSeq2NodeProps = {
  key?: string,
  seq?: Array<{value: number, time: number}>,
  duration: number,
  path: string,
  stretch?: number,
  shift?: number,
};

export function sampleseq2(props: SampleSeq2NodeProps, time: ElemNode): NodeRepr_t {
  return createNode("sampleseq2", props, [resolve(time)]);
}

// Pole node
export function pole(p: ElemNode, x: ElemNode): NodeRepr_t;
export function pole(props: OptionalKeyProps, p: ElemNode, x: ElemNode): NodeRepr_t;
export function pole(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("pole", {}, [resolve(a), resolve(b)]);
  }

  return createNode("pole", a, [resolve(b), resolve(c)]);
}

// Env node
export function env(atkPole: ElemNode, relPole: ElemNode, x: ElemNode): NodeRepr_t;
export function env(props: OptionalKeyProps, atkPole: ElemNode, relPole: ElemNode, x: ElemNode): NodeRepr_t;
export function env(a, b, c, d?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("env", {}, [resolve(a), resolve(b), resolve(c)]);
  }

  return createNode("env", a, [resolve(b), resolve(c), resolve(d)]);
}

// Single sample delay node
export function z(x: ElemNode): NodeRepr_t;
export function z(props: OptionalKeyProps, x: ElemNode): NodeRepr_t;
export function z(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("z", {}, [resolve(a)]);
  }

  return createNode("z", a, [resolve(b)]);
}

// Variable delay node
type DelayNodeProps = {
  key?: string,
  size: number,
};

export function delay(
  props: DelayNodeProps,
  len: ElemNode,
  fb: ElemNode,
  x : ElemNode,
): NodeRepr_t;

export function delay(a, b, c, d) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("delay", {}, [resolve(a), resolve(b), resolve(c)]);
  }

  return createNode("delay", a, [resolve(b), resolve(c), resolve(d)]);
}

export function sdelay(props: DelayNodeProps, x: ElemNode): NodeRepr_t {
  return createNode("sdelay", props, [resolve(x)]);
}

// Multimode1p
export function prewarp(fc: ElemNode): NodeRepr_t {
  return createNode("prewarp", {}, [fc]);
}

export function mm1p(
  fc: ElemNode,
  x: ElemNode,
): NodeRepr_t;

export function mm1p(
  props: {
    key?: string,
    mode?: string,
  },
  fc: ElemNode,
  x: ElemNode,
): NodeRepr_t;

export function mm1p(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("mm1p", {}, [
      resolve(a),
      resolve(b),
    ]);
  }

  return createNode("mm1p", a, [
    resolve(b),
    resolve(c),
  ]);
}

// SVF
export function svf(
  fc: ElemNode,
  q: ElemNode,
  x: ElemNode,
): NodeRepr_t;

export function svf(
  props: {
    key?: string,
    mode?: string,
  },
  fc: ElemNode,
  q: ElemNode,
  x: ElemNode,
): NodeRepr_t;

export function svf(a, b, c, d?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("svf", {}, [
      resolve(a),
      resolve(b),
      resolve(c),
    ]);
  }

  return createNode("svf", a, [
    resolve(b),
    resolve(c),
    resolve(d),
  ]);
}

export function svfshelf(
  fc: ElemNode,
  q: ElemNode,
  gainDecibels: ElemNode,
  x: ElemNode,
): NodeRepr_t;

export function svfshelf(
  props: {
    key?: string,
    mode?: string,
  },
  fc: ElemNode,
  q: ElemNode,
  gainDecibels: ElemNode,
  x: ElemNode,
): NodeRepr_t;

export function svfshelf(a, b, c, d, e?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("svfshelf", {}, [
      resolve(a),
      resolve(b),
      resolve(c),
      resolve(d),
    ]);
  }

  return createNode("svfshelf", a, [
    resolve(b),
    resolve(c),
    resolve(d),
    resolve(e),
  ]);
}

// Biquad node
export function biquad(
  b0: ElemNode,
  b1: ElemNode,
  b2: ElemNode,
  a1: ElemNode,
  a2: ElemNode,
  x: ElemNode,
): NodeRepr_t;

export function biquad(
  props: OptionalKeyProps,
  b0: ElemNode,
  b1: ElemNode,
  b2: ElemNode,
  a1: ElemNode,
  a2: ElemNode,
  x: ElemNode,
): NodeRepr_t;

export function biquad(a, b, c, d, e, f, g?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("biquad", {}, [
      resolve(a),
      resolve(b),
      resolve(c),
      resolve(d),
      resolve(e),
      resolve(f),
    ]);
  }

  return createNode("biquad", a, [
    resolve(b),
    resolve(c),
    resolve(d),
    resolve(e),
    resolve(f),
    resolve(g),
  ]);
}

// Feedback nodes
type TapInNodeProps = {
  name: string,
};

type TapOutNodeProps = {
  name: string,
};

export function tapIn(props: TapInNodeProps): NodeRepr_t {
  return createNode("tapIn", props, []);
}

export function tapOut(props: TapOutNodeProps, x: ElemNode): NodeRepr_t {
  return createNode("tapOut", props, [resolve(x)]);
}

// Meter node
type MeterNodeProps = {
  key?: string,
  name?: string,
};

export function meter(x: ElemNode): NodeRepr_t;
export function meter(props: MeterNodeProps, x: ElemNode): NodeRepr_t;
export function meter(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("meter", {}, [resolve(a)]);
  }

  return createNode("meter", a, [resolve(b)]);
}

// Snapshot node
type SnapshotNodeProps = {
  key?: string,
  name?: string,
};

export function snapshot(trigger: ElemNode, x: ElemNode): NodeRepr_t;
export function snapshot(props: SnapshotNodeProps, trigger: ElemNode, x: ElemNode): NodeRepr_t;
export function snapshot(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("snapshot", {}, [resolve(a), resolve(b)]);
  }

  return createNode("snapshot", a, [resolve(b), resolve(c)]);
}

// Scope node
type ScopeNodeProps = {
  key?: string,
  name?: string,
  size?: number,
  channels?: number,
};

export function scope(...args : Array<ElemNode>): NodeRepr_t;
export function scope(props: ScopeNodeProps, ...args : Array<ElemNode>): NodeRepr_t;
export function scope(a, ...bs) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("scope", {}, [a, ...bs].map(resolve));
  }

  return createNode("scope", a, bs.map(resolve));
}

// FFT node
type FFTNodeProps = {
  key?: string,
  name?: string,
  size?: number,
};

export function fft(x: ElemNode): NodeRepr_t;
export function fft(props: FFTNodeProps, x: ElemNode): NodeRepr_t;
export function fft(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("fft", {}, [resolve(a)]);
  }

  return createNode("fft", a, [resolve(b)]);
}

// Capture node
type CaptureNodeProps = {
  key?: string,
  size?: number,
};

export function capture(g: ElemNode, x: ElemNode): NodeRepr_t;
export function capture(props: CaptureNodeProps, g: ElemNode, x: ElemNode): NodeRepr_t;
export function capture(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("capture", {}, [resolve(a), resolve(b)]);
  }

  return createNode("capture", a, [resolve(b), resolve(c)]);
}
