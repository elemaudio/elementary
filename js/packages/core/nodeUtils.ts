import { create, isNode as NodeRepr_isNode } from './src/NodeRepr.gen';
import type { t as NodeRepr_t } from './src/NodeRepr.gen';

import invariant from 'invariant';


export type ElemNode = NodeRepr_t | number;
export type { NodeRepr_t };

export function resolve(n : ElemNode): NodeRepr_t {
  if (typeof n === 'number')
    return create("const", {value: n}, []);

  invariant(isNode(n), `Whoops, expecting a Node type here! Got: ${typeof n}`);
  return n;
}

export function isNode(n: unknown): n is NodeRepr_t {
  // We cannot pass `unknown` type to the underlying method generated from ReScript,
  // but we'd like to keep the signature of this method's API to be more semantically correct (use `unknown` instead of `any`).
  // That's why we're using "@ts-expect-error" here.
  // Once this resolved, the TS error pops up and we can remove it.
  // @ts-expect-error
  return NodeRepr_isNode(n);
}

export function createNode(
  kind: string,
  props,
  children: Array<ElemNode>
): NodeRepr_t {
  return create(kind, props, children.map(resolve));
}

// Utility function for addressing multiple output channels from a given graph node
export function unpack(node: NodeRepr_t, numChannels: number): Array<NodeRepr_t> {
  return Array.from({length: numChannels}, (v, i) => {
    return {
      ...node,
      outputChannel: i,
    };
  });
}
