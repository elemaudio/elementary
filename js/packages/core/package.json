{
  "name": "@elemaudio/core",
  "version": "2.0.2",
  "type": "module",
  "description": "Official Elementary Audio core package",
  "keywords": [
    "audio",
    "dsp",
    "signal",
    "processing",
    "functional",
    "declarative",
    "webaudio"
  ],
  "author": "Nick Thompson <nick@elementary.audio>",
  "homepage": "https://www.elementary.audio",
  "license": "MIT",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./package.json": "./package.json"
  },
  "files": [
    "dist/index.cjs",
    "dist/index.js",
    "dist/index.d.ts",
    "README.md",
    "LICENSE.md"
  ],
  "scripts": {
    "re:build": "rescript",
    "re:start": "rescript build -w",
    "clean": "rm -rf ./dist/",
    "prebuild": "npm run clean",
    "build": "npm run re:build && tsup index.ts --format cjs,esm --dts",
    "test": "jest",
    "snaps": "jest --updateSnapshot"
  },
  "jest": {
    "verbose": true
  },
  "devDependencies": {
    "@babel/preset-env": "^7.16.11",
    "gentype": "^4.5.0",
    "jest": "^27.5.1",
    "rescript": "^10.0.0",
    "tsup": "^6.0.1"
  },
  "dependencies": {
    "@types/events": "^3.0.0",
    "events": "^3.3.0",
    "invariant": "^2.2.4",
    "shallowequal": "^1.1.0"
  }
}
