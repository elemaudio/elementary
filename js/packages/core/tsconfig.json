{
  "compilerOptions": {
    "allowJs": true,
    "declaration": true,
    "target": "es6",
    "module": "es6",
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true
  },
  "files": [
    "index.ts",
  ]
}
