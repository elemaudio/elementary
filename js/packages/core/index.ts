import {
  renderWithDelegate,
  stepGarbageCollector,
} from './src/Reconciler.gen';

import { updateNodeProps } from './src/Hash';

import {
  createNode,
  unpack,
  isNode,
  resolve,
  NodeRepr_t,
} from './nodeUtils';


import * as co from './lib/core';
import * as dy from './lib/dynamics';
import * as en from './lib/envelopes';
import * as ma from './lib/math';
import * as mc from './lib/mc';
import * as fi from './lib/filters';
import * as os from './lib/oscillators';
import * as si from './lib/signals';

export type { ElemNode, NodeRepr_t } from './nodeUtils';
export { default as EventEmitter } from './src/Events';


const stdlib = {
  ...co,
  ...dy,
  ...en,
  ...fi,
  ...ma,
  ...os,
  ...si,
  mc,
  // Aliases for reserved keyword conflicts
  "const": co.constant,
  "in": ma.identity,
};

const InstructionTypes = {
  CREATE_NODE: 0,
  DELETE_NODE: 1,
  APPEND_CHILD: 2,
  SET_PROPERTY: 3,
  ACTIVATE_ROOTS: 4,
  COMMIT_UPDATES: 5,
};

// A default render delegate which batches instruction sets while recording
// stats about the render pass.
class Delegate {
  public nodesAdded: number;
  public nodesRemoved: number;
  public edgesAdded: number;
  public propsWritten: number;

  public nodeMap: Map<number, any>;

  private currentActiveRoots: Set<number>;
  private terminalGeneration: number;
  private batch: any;

  constructor() {
    this.nodeMap = new Map();
    this.currentActiveRoots = new Set();

    this.clear();
  }

  clear() {
    this.nodesAdded = 0;
    this.nodesRemoved = 0;
    this.edgesAdded = 0;
    this.propsWritten = 0;

    this.batch = {
      createNode: [],
      appendChild: [],
      setProperty: [],
      activateRoots: [],
      commitUpdates: [],
      deleteNode: []
    };
  }

  getNodeMap() { return this.nodeMap; }
  getTerminalGeneration() { return 4; }

  createNode(hash, type) {
    this.nodesAdded++;
    this.batch.createNode.push([InstructionTypes.CREATE_NODE, hash, type]);
  }

  deleteNode(hash) {
    this.nodesRemoved++;
    this.batch.deleteNode.push([InstructionTypes.DELETE_NODE, hash]);
  }

  appendChild(parentHash, childHash, childOutputChannel) {
    this.edgesAdded++;
    this.batch.appendChild.push([InstructionTypes.APPEND_CHILD, parentHash, childHash, childOutputChannel]);
  }

  setProperty(hash, key, value) {
    this.propsWritten++;
    this.batch.setProperty.push([InstructionTypes.SET_PROPERTY, hash, key, value]);
  }

  activateRoots(roots) {
    // If we're asked to activate exactly the roots that are already active,
    // no need to push the instruction. We need the length/size check though
    // because it may be that we're asked to activate a subset of the current
    // active roots, in which case we need the instruction to prompt the engine
    // to deactivate the now excluded roots.
    let alreadyActive = roots.length === this.currentActiveRoots.size &&
      roots.every((root) => this.currentActiveRoots.has(root));

    if (!alreadyActive) {
      this.batch.activateRoots.push([InstructionTypes.ACTIVATE_ROOTS, roots]);
      this.currentActiveRoots = new Set(roots);
    }
  }

  commitUpdates() {
    this.batch.commitUpdates.push([InstructionTypes.COMMIT_UPDATES]);
  }

  getPackedInstructions() {
    return [
      ...this.batch.createNode,
      ...this.batch.appendChild,
      ...this.batch.setProperty,
      ...this.batch.activateRoots,
      ...this.batch.commitUpdates,
      ...this.batch.deleteNode,
    ];
  }
}

// A quick shim for platforms which do not support the `performance` global
function now() {
  if (typeof performance === 'undefined') {
    return Date.now();
  }

  return performance.now();
}

// A generic Renderer just wraps the above Delegate to provide a simple
// render root.
//
// Most applications will just need this generic Renderer with a callback for
// marshaling instructions over to the engine, wherever that may be.
//
// Applications which need more fine grained control can write their own Renderer
// or their own Delegate.
class Renderer {
  private _delegate: Delegate;
  private _sendMessage: Function;
  private _nextRefId: number;

  constructor(sendMessage) {
    this._delegate = new Delegate();
    this._sendMessage = sendMessage;
    this._nextRefId = 0;
  }

  // A method for creating "refs," which looks the same as the function for creating
  // nodes but captures the context of the Renderer instance to provide scoped property
  // updates to the ref without incurring a full graph construction and reconciliation pass.
  //
  // Example:
  //  let [cutoffFreq, setCutoffFreq] = createRef("const", {value: 440}, []);
  //
  //  // Render a ref just the same as you would any other node
  //  core.render(el.lowpass(cutoffFreq, 1, el.in({channel: 0})));
  //
  //  // Subsequent property changes can be made through the property setter returned
  //  // from the call to createRef
  //  setCutoffFreq({ value: 440 });
  //
  // Note: refs should only be rendered by the Renderer instance from which they were created.
  // In other words, don't share refs between different renderer instances.
  createRef(kind, props, children) {
    let key = `__refKey:${this._nextRefId++}`;
    let node = createNode(kind, Object.assign({key}, props), children);

    let setter = (newProps) => {
      if (!this._delegate.nodeMap.has(node.hash)) {
        throw new Error('Cannot update a ref that has not been mounted; make sure you render your node first')
      }

      const nodeMapCopy = this._delegate.nodeMap.get(node.hash);

      this._delegate.clear();
      updateNodeProps(this._delegate, node.hash, nodeMapCopy.props, newProps);
      this._delegate.commitUpdates();

      // Invoke message passing
      const instructions = this._delegate.getPackedInstructions();

      return Promise.resolve(this._sendMessage(instructions));
    };

    return [node, setter];
  }

  render(...args) {
    const t0 = now();

    this._delegate.clear();
    renderWithDelegate(this._delegate as any, args.map(resolve));

    const t1 = now();

    // Invoke message passing
    const instructions = this._delegate.getPackedInstructions();

    return Promise.resolve(this._sendMessage(instructions)).then((result) => {
      // Pack render stats with the result of the message passing
      return {
        result,
        nodesAdded: this._delegate.nodesAdded,
        edgesAdded: this._delegate.edgesAdded,
        propsWritten: this._delegate.propsWritten,
        elapsedTimeMs: t1 - t0,
      };
    });
  }
}

export {
  Delegate,
  Renderer,
  createNode,
  unpack,
  isNode,
  resolve,
  renderWithDelegate,
  stepGarbageCollector,
  stdlib,
  stdlib as el,
};
