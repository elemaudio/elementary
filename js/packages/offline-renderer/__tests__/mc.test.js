import OfflineRenderer from '../index';
import { el } from '@elemaudio/core';


test('mc table', async function() {
  let core = new OfflineRenderer();

  await core.initialize({
    numInputChannels: 0,
    numOutputChannels: 1,
    virtualFileSystem: {
      '/v/ones': Float32Array.from([1, 1, 1]),
      '/v/stereo': [
        Float32Array.from([27, 27, 27]),
        Float32Array.from([15, 15, 15]),
      ],
    },
  });

  // Graph
  await core.render(el.add(...el.mc.table({path: '/v/stereo', channels: 2}, 0)));

  // Ten blocks of data
  let inps = [];
  let outs = [new Float32Array(512 * 10)];

  // Get past the fade-in
  core.process(inps, outs);

  // Process another small block
  inps = [];
  outs = [new Float32Array(16)];

  core.process(inps, outs);

  expect(outs[0]).toMatchSnapshot();
});

test('mc sampleseq', async function() {
  let core = new OfflineRenderer();

  await core.initialize({
    numInputChannels: 0,
    numOutputChannels: 1,
    virtualFileSystem: {
      '/v/stereo': [
        Float32Array.from(Array.from({length: 128}).fill(27)),
        Float32Array.from(Array.from({length: 128}).fill(15)),
      ],
    },
  });

  let [time, setTimeProps] = core.createRef("const", {value: 0}, []);

  core.render(
    el.add(...el.mc.sampleseq({
      path: '/v/stereo',
      channels: 2,
      duration: 128,
      seq: [
        { time: 0, value: 0 },
        { time: 128, value: 1 },
        { time: 256, value: 0 },
        { time: 512, value: 1 },
      ]
    }, time))
  );

  // Ten blocks of data to get past the root node fade
  let inps = [];
  let outs = [new Float32Array(10 * 512)];

  // Get past the fade-in
  core.process(inps, outs);

  // Now we expect to see zeros because we're fixed at time 0
  outs = [new Float32Array(32)];
  core.process(inps, outs);
  expect(outs[0]).toMatchSnapshot();

  // Jump in time
  setTimeProps({value: 520});

  // Spin for a few blocks and we should see the gain fade resolve and
  // emit the constant sum of the two channels
  for (let i = 0; i < 10; ++i) {
    core.process(inps, outs);
  }

  expect(outs[0]).toMatchSnapshot();
});
