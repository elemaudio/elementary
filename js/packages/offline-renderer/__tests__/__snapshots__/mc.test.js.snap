// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mc sampleseq 1`] = `
Float32Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`mc sampleseq 2`] = `
Float32Array [
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
]
`;

exports[`mc table 1`] = `
Float32Array [
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
]
`;
